src/Npgsql/NpgsqlDatabaseInfo.cs

```diff
@@ -221,13 +221,17 @@ public static void RegisterFactory(INpgsqlDatabaseInfoFactory factory)
             if (factory == null)
                 throw new ArgumentNullException(nameof(factory));
 
-            Factories.Insert(0, factory);
+            var factories = new List<INpgsqlDatabaseInfoFactory>(Factories);
+            factories.Insert(0, factory);
+            Factories = factories;
+
             Cache.Clear();
         }
 
         internal static async Task<NpgsqlDatabaseInfo> Load(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async)
         {
-            foreach (var factory in Factories)
+            var factories = Factories;
+            foreach (var factory in factories)
```

Comment:
For non array collections that isn't required since only arrays has a special treating of `foreach` which is translated to `for` to eliminate enumerator and bound checks. For other cases, `foreach` is untouched.